File: web/src/lib/manifest.ts
/**
 * Types and loaders for the model asset bundle (spec §3.2-3.4).
 *
 * The viewer is manifest-driven: it reads scenarios, variables, paths, value
 * ranges, grid sizes, and colormaps from here and NEVER hard-codes them
 * (CLAUDE.md cardinal rule #3). The bundle is produced by model/run.py (Phase 3)
 * and lives under /assets/model/.
 */

export interface VariableManifest {
  /** Frame path template, e.g. "summer-orbit/temperature/frame_{i:04d}.png". */
  frames: string;
  /** Packed Float32 probe blob (the scalar field), one per (scenario, variable). */
  probe_blob: string;
  /** Probe grid dimensions [w, h]. */
  grid_size: [number, number];
  /** Fixed value range [min, max] used for the baked colormap. */
  value_range: [number, number];
  units: string;
  colormap: string;
  /** Wind only: packed Float32 u/v component blobs for particle advection. */
  components?: { u: string; v: string };
}

export interface ScenarioManifest {
  id: string;
  label: string;
  kind?: "solved" | "reference";
  source?: string;
  timesteps: number;
  time_labels?: string[];
  variables: Record<string, VariableManifest>;
  archive?: string | null;
}

export interface DiscManifest {
  radius_km: number;
  center: string;
}

export interface Manifest {
  scenarios: ScenarioManifest[];
  disc: DiscManifest;
  display_size?: number;
}

export const ASSET_BASE = "/assets/model/";
export const DEFAULT_MANIFEST_URL = ASSET_BASE + "manifest.json";

/** Resolve a manifest-relative path to a URL under the asset base. */
export function assetUrl(rel: string): string {
  return ASSET_BASE + rel;
}

/** Fetch and parse the manifest. Throws if it is missing or malformed. */
export async function loadManifest(
  url: string = DEFAULT_MANIFEST_URL,
): Promise<Manifest> {
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`failed to load manifest (${res.status}) from ${url}`);
  }
  return (await res.json()) as Manifest;
}

/** Expand a frame template for timestep `i` ("{i}" / zero-padded "{i:0Nd}"). */
export function frameUrl(template: string, i: number): string {
  return template.replace(/\{i(?::0(\d+)d)?\}/g, (_full, width?: string) =>
    width ? String(i).padStart(Number(width), "0") : String(i),
  );
}

/** Full URL of a display PNG for (variable template, timestep). */
export function frameAssetUrl(template: string, i: number): string {
  return assetUrl(frameUrl(template, i));
}

// One in-flight/parsed promise per probe blob path, so each is fetched once.
const probeCache = new Map<string, Promise<Float32Array>>();

/** Fetch a packed Float32 probe blob (cached across the session). */
export function loadProbeBlob(rel: string): Promise<Float32Array> {
  let pending = probeCache.get(rel);
  if (!pending) {
    pending = fetch(assetUrl(rel)).then(async (res) => {
      if (!res.ok) throw new Error(`probe ${rel}: ${res.status}`);
      return new Float32Array(await res.arrayBuffer());
    });
    probeCache.set(rel, pending);
  }
  return pending;
}

/** A view onto timestep `i` of a packed probe blob (length w*h, C-order). */
export function sliceFrame(
  blob: Float32Array,
  i: number,
  w: number,
  h: number,
): Float32Array {
  const n = w * h;
  return blob.subarray(i * n, (i + 1) * n);
}

File: model/render/manifest.py
"""Write and validate the asset-bundle ``manifest.json`` (spec §3.4).

A single manifest at the asset root describes every scenario, variable, frame
path template, probe-blob path, grid size, value range, units, and colormap,
plus the disc geometry. The viewer reads ONLY the manifest and never hard-codes
paths, frame counts, or value ranges (CLAUDE.md, cardinal rule #3).

Frame templates use the ``{i:04d}`` form, e.g.
``"summer-orbit/temperature/frame_{i:04d}.png"``.
"""

from __future__ import annotations

import json
from pathlib import Path

_FLOAT32_BYTES = 4


def write_manifest(
    scenarios: list[dict],
    out_path: Path,
    *,
    disc_radius_km: float,
    display_size: int,
) -> None:
    """Write ``manifest.json`` describing the produced asset bundle."""
    manifest = {
        "scenarios": scenarios,
        "disc": {"radius_km": disc_radius_km, "center": "north_pole"},
        "display_size": display_size,
        "format": "display-png + float32-probe-bin (ARCHITECTURE.md §3.2)",
    }
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(manifest, indent=2) + "\n")


def _check_blob(
    base_dir: Path, rel: str, timesteps: int, w: int, h: int, issues: list[str]
) -> None:
    blob = base_dir / rel
    if not blob.exists():
        issues.append(f"missing probe blob: {rel}")
        return
    expected = timesteps * w * h * _FLOAT32_BYTES
    actual = blob.stat().st_size
    if actual != expected:
        issues.append(f"{rel}: size {actual} != expected {expected} ({timesteps}x{w}x{h} f32)")


def validate_manifest(manifest: dict, base_dir: Path) -> list[str]:
    """Return a list of problems with the bundle under ``base_dir`` (empty = valid).

    Checks that every frame PNG exists and every probe blob has exactly
    ``timesteps * w * h * 4`` bytes (and likewise for wind component blobs).
    """
    issues: list[str] = []
    if "disc" not in manifest or "scenarios" not in manifest:
        issues.append("manifest missing 'disc' or 'scenarios'")
        return issues

    for sc in manifest["scenarios"]:
        steps = sc["timesteps"]
        for var, meta in sc["variables"].items():
            w, h = meta["grid_size"]
            for i in range(steps):
                frame = base_dir / meta["frames"].format(i=i)
                if not frame.exists():
                    issues.append(f"{sc['id']}/{var}: missing frame {frame.name}")
                    break  # one report per variable is enough
            _check_blob(base_dir, meta["probe_blob"], steps, w, h, issues)
            for comp_rel in meta.get("components", {}).values():
                _check_blob(base_dir, comp_rel, steps, w, h, issues)
    return issues
